util/src/snowflake.rs

```diff
@@ -0,0 +1,158 @@
+//! A module providing the Snowflake trait for defining extractable information from a Discord
+//! Snowflake.
+
+/// Snowflake is a trait for defining extractable information from a Snowflake. A Snowflake is a
+/// u64 generated by Discord to uniquely identify a resource.
+pub trait Snowflake {
+    /// Returns the u64 backing the Snowflake.
+    fn id(&self) -> u64;
+
+    /// The Unix epoch of the Snowflake in milliseconds, indicating when it was generated.
+    ///
+    /// 22 bits
+    ///
+    /// # Examples
+    ///
+    /// See when a user was created using [`chrono`](https://docs.rs/chrono):
+    ///
+    /// ```rust
+    /// use chrono::{Utc, TimeZone};
+    /// use twilight_util::snowflake::Snowflake;
+    /// use twilight_model::id::UserId;
+    ///
+    /// let id = UserId(105484726235607040);
+    ///
+    /// assert_eq!(
+    ///     "2015-10-19T01:58:38.546+00:00",
+    ///     Utc.timestamp_millis(id.timestamp()).to_rfc3339()
+    /// );
+    /// ```
+    ///
+    /// See when a user was created using [`time`](https://docs.rs/time):
+    ///
+    /// ```rust
+    /// use time::{Duration, Format, OffsetDateTime};
+    /// use twilight_util::snowflake::Snowflake;
+    /// use twilight_model::id::UserId;
+    ///
+    /// let id = UserId(105484726235607040);
+    /// let dur = Duration::milliseconds(id.timestamp());
+    /// // Or use seconds, at the cost of lost precision.
+    /// let ts = OffsetDateTime::from_unix_timestamp_nanos(dur.whole_nanoseconds());
+    ///
+    /// assert_eq!("2015-10-19T01:58:38+00:00", ts.format(Format::Rfc3339));
+    /// ```
+    fn timestamp(&self) -> i64 {
+        // Discord's custom epoch, the unix time in milliseconds for the first second of 2015.
+        const DISCORD_EPOCH: u64 = 1420070400000;
```

Comment:
We should make this a constant in the trait

Reply:
Adding this to the trait prevents it from being turned into an object unfortunately.